### ISSUE TITLE
`ProcessWorker` doesn't emit flow run heartbeats

### ISSUE THREAD

**Issue by ihor-ramskyi-globallogic** [PrefectHQ/prefect#16515]
### Bug summary

I followed the [instruction](https://github.com/PrefectHQ/prefect/releases/tag/3.1.8) to enable automation for crashing zombie flows.
When I tried it on flow that was deployed using `.serve()`, it generated heartbeats, and when I tried deploying same flow using `prefect --no-prompt deploy --prefect-file prefect.yaml -n Zombie`, it didn't
![image](https://github.com/user-attachments/assets/1a6d5cd4-3c29-43b4-aeec-21bc361ece6f)
![image](https://github.com/user-attachments/assets/544be0e8-040a-441f-9f64-5c956be07e26)

But even with flow that was generating heartbeats successfully, automation from the same instruction didn't work. My procedure was:
1. Serve the flow
2. Run it
3. Wait for the first heartbeat to happen
4. Stop serving process.

Result: flow continues to run as if automation wasn't there.

Screenshots:
![image](https://github.com/user-attachments/assets/c998e795-d64c-42eb-bc46-4aeef3844958)
![image](https://github.com/user-attachments/assets/d2a6ae80-8407-4642-a2cc-384c7658288f)
![image](https://github.com/user-attachments/assets/ea1012e5-6ea0-429e-aa40-3f9caac68cc8)

<details>
<summary>Setup description:</summary>

Flow code:

```
import time

from prefect import flow, task, get_run_logger

@task
def task_a():
    logger = get_run_logger()
    logger.info("Hello")
    time.sleep(5)

@flow
def flow_a():
    while True:
        task_a.submit().result()

if __name__ == "__main__":
    flow_a.serve()
```

`prefect.yaml`:

```
name: Zombie

work-queue-default: &default
  work_pool:
    name: work_pool_1
    work_queue_name: default
    job_variables:
      stream_output: true


deployments:
- name: Zombie
  entrypoint: zombie.py:flow_a
  <<: *default
```

`prefect config view`:

```
🚀 you are connected to:
http://127.0.0.1:4200
PREFECT_PROFILE='default'
PREFECT_API_KEY='********' (from profile)
PREFECT_API_URL='http://127.0.0.1:4200/api' (from profile)
PREFECT_LOCAL_STORAGE_PATH='C:\Users\ihor.ramskyi\PycharmProjects\Prefect_Tutorial\local_storage' (from profile)
PREFECT_RUNNER_HEARTBEAT_FREQUENCY='30' (from profile)
```

Worker was started using command `prefect worker start -p work_pool_1 -t process -q default`
Flow was deployed using command `prefect --no-prompt deploy --prefect-file prefect.yaml -n Zombie`
</details>

### Version info

```Text
Version:             3.1.8
API version:         0.8.4
Python version:      3.10.11
Git commit:          53a83ebc
Built:               Tue, Dec 17, 2024 10:20 AM
OS/Arch:             win32/AMD64
Profile:             default
Server type:         server
Pydantic version:    2.7.1
```

**Comment by desertaxle** [comment]
Thanks for the issue @ihor-ramskyi-globallogic! My hunch is that the heartbeat setting is getting set in your execution environment. I think the easiest way to do that is to add a `prefect.toml` file in the root of your project that looks like this:
```toml
runner.heartbeat_frequency = 30
```
Once that file is created, you can redeploy your flow, and you should see heartbeats coming from flow runs for that deployment.

Let me know if you run into any more issues with this! If creating a `prefect.toml` works for you, we should update the docs to include that step.

**Comment by dsmfreire** [comment]
@desertaxle is the heartbeat feature only available when we deploy flows through .serve()? I see that it is implemented in the Runner, which is exclusively used in this mode (at least according to its docstring).

**Comment by desertaxle** [comment]
Great question @dsmfreire! When a worker creates infrastructure for a flow to run in (e.g. a Kubernetes job), a Runner is started up to monitor the execution of the flow run. So as long as you set the heartbeat frequency setting, you can also get heartbeat events for flows deployed to work pools!

**Comment by ihor-ramskyi-globallogic** [comment]
> Thanks for the issue @ihor-ramskyi-globallogic! My hunch is that the heartbeat setting isn't getting set in your execution environment. I think the easiest way to do that is to add a `prefect.toml` file in the root of your project that looks like this:
> 
> ```toml
> runner.heartbeat_frequency = 30
> ```
> 
> Once that file is created, you can redeploy your flow, and you should see heartbeats coming from flow runs for that deployment.
> 
> Let me know if you run into any more issues with this! If creating a `prefect.toml` works for you, we should update the docs to include that step.

I tried, but it doesn't work.
![image](https://github.com/user-attachments/assets/853a091e-f047-450d-af84-e571c8afab16)
![image](https://github.com/user-attachments/assets/947a8f25-0073-4137-89c4-f723050d4f23)

<details>
<summary>Setup</summary>

My toml:

```toml
runner.heartbeat_frequency = 30
logging.level = "DEBUG"
```

My code:

```python
import time
import subprocess

from prefect import flow, task, get_run_logger


@task
def task_a():
    logger = get_run_logger()
    logger.info("Oui")
    time.sleep(5)


@flow(name="Zombie")
def zombie():
    logger = get_run_logger()
    logger.debug("toml works")
    logger.info("Version: " + subprocess.run("prefect --version", check=True, text=True, shell=True, capture_output=True).stdout)
    while True:
        task_a.submit().result()


if __name__ == "__main__":
    zombie.serve()

```
</details>

**Comment by andreas-ntonas** [comment]
> Great question [@dsmfreire](https://github.com/dsmfreire)! When a worker creates infrastructure for a flow to run in (e.g. a Kubernetes job), a Runner is started up to monitor the execution of the flow run. So as long as you set the heartbeat frequency setting, you can also get heartbeat events for flows deployed to work pools!

This functionality doesn't seem to work for the process worker type for me. Are we sure that a Runner is created for all worker types? 

By a brief look in the code it seems to me that the process worker [uses a different set of tasks](https://github.com/PrefectHQ/prefect/blob/main/src/prefect/workers/process.py#L186-L217) without an `_emit_flow_run_heartbeats` job that the Runner is using [in similar code](https://github.com/PrefectHQ/prefect/blob/main/src/prefect/runner/runner.py#L641-L660).

**Comment by desertaxle** [comment]
Yup, that's a great catch @andreas-ntonas! The process worker hasn't been updated to emit flow run heartbeats mainly because the process worker doesn't keep track of stuff that it's running in the same way the `Runner` does.

The `Runner` started out as a modified copy of the `ProcessWorker` class and has evolved to include some features that the `ProcessWorker` doesn't have (like flow run heartbeats), but they perform very similar roles (polling for and executing flow runs). I'd like to rationalize the `Runner` and `ProcessWorker` so there's less overlap in functionality. The `ProcessWorker` can be the poller, while the `Runner` is only responsible for executing single flow runs.

In the meantime, you can use a `Runner` directly if you want to execute flow runs in subprocesses with flow run heartbeats, but we'll need to spend some time to get the `ProcessWorker` into a similar state. Let me know if you're interested in helping out with that at all!
